src/index/ivf_raft/ivf_raft.cuh

```diff
@@ -568,63 +532,24 @@ class RaftIvfIndexNode : public IndexNode {
-                            ids_gpu_before.view(), dis_gpu_before.view());
-
-                        ids_flush<<<(rows * k2 + 1023) / 1024, 1024, 0, stream.value()>>>(ids_gpu_before.data_handle(),
-                                                                                          rows * k2);
-                        auto ids_block = raft::make_device_vector<int32_t, std::int64_t>(*res_, rows * k2);
-                        filter<<<(rows * k2 + 1023) / 1024, 1024, 0, stream.value()>>>(
-                            bs_gpu.data_handle(), ids_gpu_before.data_handle(), ids_block.data_handle(), rows * k2);
-                        auto pre_ids_block = raft::make_device_vector<int32_t, std::int64_t>(*res_, rows * k2);
-                        RAFT_CUDA_TRY(cudaMemcpyAsync(pre_ids_block.data_handle(), ids_block.data_handle(),
-                                                      rows * k2 * sizeof(int32_t), cudaMemcpyDefault, stream.value()));
-                        prescan<<<dim3(1, rows), k2 / 2, sizeof(int32_t) * k2, stream.value()>>>(
-                            pre_ids_block.data_handle(), ids_block.data_handle(), k2);
-                        write_back<<<dim3(1, rows), k1, sizeof(int32_t) * k2, stream.value()>>>(
-                            pre_ids_block.data_handle(), ids_gpu_before.data_handle(), dis_gpu_before.data_handle(),
-                            ids_gpu.data_handle(), dis_gpu.data_handle(), k1, k2);
-
-                        std::int64_t is_fine = 0;
-                        RAFT_CUDA_TRY(cudaMemcpyAsync(&is_fine, ids_gpu_before.data_handle(), sizeof(std::int64_t),
-                                                      cudaMemcpyDefault, stream.value()));
-                        stream.synchronize();
-                        if (is_fine != -2)
-                            break;
-                        k2 = k2 << 1;
-                    }
-                }
+                gpu_results = RawSearch(*res_, raft::make_const_mdspan(data_gpu.view()), search_params, ivf_raft_cfg.k,
+                                        ivf_raft_cfg.k, gpu_bitset.view());
             } else {
                 static_assert(std::is_same_v<detail::raft_ivf_flat_index, T>);
             }
-            RAFT_CUDA_TRY(cudaMemcpyAsync(ids.get(), ids_gpu.data_handle(), ids_gpu.size() * sizeof(std::int64_t),
-                                          cudaMemcpyDefault, stream.value()));
-            RAFT_CUDA_TRY(cudaMemcpyAsync(dis.get(), dis_gpu.data_handle(), dis_gpu.size() * sizeof(float),
-                                          cudaMemcpyDefault, stream.value()));
-            stream.synchronize();
+
+            if (gpu_results.k() != ivf_raft_cfg.k) {
+                auto new_gpu_results = raft_detail::raft_results{*res_, gpu_results.rows(), ivf_raft_cfg.k};
+                raft_detail::slice<<<1024, 256, 0, res_->get_stream().value()>>>(
```

Comment:
Just curious- where did the choice of 1024 for the number of blocks come from? This could probably be chosen based on the number of SMs.

Reply:
Just a safe default for all architectures we're building for. Given the time constraints, I didn't want to do the necessary device querying to get the optimal configuration. I don't expect this to be a bottleneck anyway just yet, and there are a bunch of other performance tweaks that will probably be even higher impact. That being said, we should definitely improve on this in subsequent iterations.